bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -2576,114 +2656,137 @@ void BCStateTran::processData(bool lastInBatch) {
+            LOG_DEBUG(logger_,
+                      "Done putting (async) blocks ["
+                          << oldFetchState_.minBlockId << "," << oldFetchState_.maxBlockId << "], new fetchState_:"
+                          << fetchState_ << KVLOG(postProcessingUpperBoundBlockId_, commitState_.upperBoundBlockId));
+          }
           if (lastInBatch || postponedSendFetchBlocksMsg_ || newSourceReplica) {
-            trySendFetchBlocksMsg(firstRequiredBlock,
-                                  nextRequiredBlock_,
-                                  0,
-                                  KVLOG(lastInBatch, postponedSendFetchBlocksMsg_, newSourceReplica));
+            trySendFetchBlocksMsg(0, KVLOG(lastInBatch, postponedSendFetchBlocksMsg_, newSourceReplica));
             break;
           }
-        } else {
+        } else {  // lastFetchedBlockIdInCycle == true
           //////////////////////////////////////////////////////////////////////////
-          // This is the last block we need in cycle or on a fetch range
+          // Last block to collect in cycle
           //////////////////////////////////////////////////////////////////////////
-          LOG_INFO(logger_, ss.str());
-          commitToChainDT_.start();
-          blocks_collected_.pause();
-          bytes_collected_.pause();
-          finalizePutblockAsync(PutBlockWaitPolicy::WAIT_ALL_JOBS, true);
+          // TODO - this block code is intermediate (phase 1). ST main thread:
+          // 1) Finalizes all blocks with WAIT_ALL_JOBS and block
+          // 2) Waits for post-processing thread to finish
+          // 3) Put the last block and performs the last post-processing by itself
 
-          ConcordAssertEQ(nextBlockIdToCommit_, nextRequiredBlock_);
-          ConcordAssert(ioContexts_.empty());
-          ConcordAssert(as_->putBlock(nextRequiredBlock_, buffer_.get(), actualBlockSize, true));
+          // TODO - uncomment and fix as part of future task
+          // blocks_collected_.pause();
+          // bytes_collected_.pause();
 
-          commitToChainDT_.pause();
-          digestOfNextRequiredBlock_.makeZero();
+          // Wait for all jobs to finish
+          finalizePutblockAsync(PutBlockWaitPolicy::WAIT_ALL_JOBS);
+
+          // At this stage we haven't yet committed the last block in cycle so we expect the next assert:
+          ConcordAssertEQ(fetchState_, commitState_);
+          ConcordAssert(ioContexts_.empty());
+          // TODO - uncomment and fix as part of future task
+          // commitToChainDT_.pause();
           clearAllPendingItemsData();
+
+          // This code is temporary - lets make things simple and wait on an infinite loop for the post-processing
+          // thread to finis
+          LOG_INFO(logger_, "Waiting for post processor thread to finish all jobs in queue...");
+          while (!postProcessingQ_->empty() || (postProcessingUpperBoundBlockId_ != maxPostprocessedBlockId_)) {
```

Comment:
Should this be conditional timed wait instead of infinite loop?

Reply:
If the code was final - I agree.
Since I'm going top changes it very soon (see comment above) - it doesn't worth the effort in my opyion.
The final code should enter a new state here - waiting for checkpooint summary messages + checking the post processing thread every 250-300 ms using the periodic timer. 
So having a conditional variable here is much more work and doesn't worth the effort.